### ISSUE TITLE
Why did you delete bi-attention code?

### ISSUE THREAD

**Issue by sunghoon014** [THUcqb/KOBE#50]
Version 2 has no bi-attention code, why?

**Comment by qibinc** [comment]
Hi @sunghoon014 ,

Thanks for your question! Indeed, bi-attention was a common approach to combine different types of inputs (e.g., https://arxiv.org/pdf/1611.01603.pdf) when this work KOBE was originally implemented. It benefits from the computation efficiency since the computation cost is only `L1 x L2`. However, with the rise of BERT and its subsequent works on large Transformer models, it's now common practice now to use a sep token to concatenate the input directly (recent DeepMind ReTro tries to bring it back but mainly for computation efficiency).

Since the purpose of v2 is to provide better code quality and deliver good practical results instead of research, we have replaced the original bi-attention with the concat approach since its implementation is more concise and the performance is similar. Similarly, we have replaced beam search with nucleus sampling (https://arxiv.org/abs/1904.09751), which is now common practice for text generation tasks like this. 

For interested developers, our bi-attention implementation can be found at https://github.com/THUDM/KOBE/blob/565b74884b4680b05ac9c406f28d130709797851/core/models/transformer.py.

Best,
